nodejs/dist/uws.js

```diff
@@ -1,98 +1,134 @@
+
+        this.server.nativeServer.close(this.nativeSocket);
+        this.server = this.nativeSocket = undefined;
     }
 }
 
-module.exports.Server = function Server(options) {
-    EventEmitter.call(this);
+class Server extends EventEmitter {
+    /**
+     * Creates a Server instance.
+     *
+     * @param {Object} options Configuration options.
+     */
+    constructor(options) {
+        super();
+        /* we currently do only support port as option */
+        this.nativeServer = new NativeServer(options.port);
 
-    /* we currently do only support port as option */
-    this.nativeServer = new NativeServer(options.port);
-    var self = this;
+        this.nativeServer.onConnection((nativeSocket) => {
+            const socket = new Socket(nativeSocket, this);
+            this.nativeServer.setData(nativeSocket, socket);
+            this.emit('connection', socket);
+        });
 
-    this.close = function () {
-        self.nativeServer.close();
+        this.nativeServer.onDisconnection((nativeSocket) => {
+            const socket = self.nativeServer.getData(nativeSocket);
+            socket.emit('close');
+            /* make sure to clear any set data */
+            self.nativeServer.setData(nativeSocket);
+        });
+
+        this.nativeServer.onMessage((nativeSocket, message, binary) => {
+            const socket = self.nativeServer.getData(nativeSocket);
+            socket.emit('message', binary ? message : message.toString());
+        });
     }
-    this.broadcast = function (message, options) {
+
+    /**
+     * Closes the server.
+     *
+     * @public
+     */
+    close() {
+        this.nativeServer.close();
+    }
+
+    /**
+     * Broadcast a message to all sockets.
+     *
+     * @param {String|Buffer} message The message to broadcast
+     * @param {Object} options Broadcast options
+     * @public
+     */
+    broadcast(message, options) {
         /* only listen to binary option */
-        this.nativeServer.broadcast(message, options.binary);
-    };
-
-    /* original connection callback */
-    this.connectionCallback = function (nativeSocket) {
-        var socket = new Socket(nativeSocket, self);
-        self.nativeServer.setData(nativeSocket, socket);
-        self.emit('connection', socket);
-    };
-
-    /* this function needs to handle the upgrade directly! */
-    this.handleUpgrade = function (request, socket, upgradeHead, callback) {
+        this.nativeServer.broadcast(message, options && options.binary || false);
+    }
+
+    /**
+     * Handles a HTTP Upgrade request.
+     *
+     * @param {http.IncomingMessage} request HTTP request
+     * @param {net.Socket} Socket between server and client
+     * @param {Buffer} upgradeHead The first packet of the upgraded stream
+     * @param {Function} callback Callback function
+     * @public
+     */
+    handleUpgrade(request, socket, upgradeHead, callback) {
+        const nativeServer = this.nativeServer;
```

Comment:
Interesting that it's faster. In my experience and what I've learned about V8 in these circumstances, is that it's usually better to leave that kind of optimization to V8, since now you may simply be reducing our available CPU registers by one without benefit (this is very evident when people think `for (var i = 0, len = arr.length; i < len; i++) {}` is faster for example than not caching the length, which is simply not the case).